File: Sorting/Merge Sort/MergeSort.java
public class MergeSort {

    // Combine function
    private static void combine(int a[], int l, int m, int u) {

        int i=l, j=m+1, k=0;
        int[] temp = new int[u-l+1];

    	// Compare lower & upper part and store in temp
    	while (i<=m && j<=u) {
    		if ( a[i] <= a[j] ) {
    			temp[k++] = a[i++];
    		} else {
    			temp[k++] = a[j++];
    		}
    	}

	    // Copy remaining lower part
	    while(i<=m) { temp[k++] = a[i++]; }

	    // Copy remaining upper part
	    while(j<=u) { temp[k++] = a[j++]; }

	    // Copy sorted temp to main array
	    for (i=l; i<=u; i++) { a[i] = temp[i-l]; }

    }

    // Merge Sorting Function
    private static void merge(int[] a, int lower, int upper) {

        if (lower < upper) {
            int middle = (lower + upper)/2;

            merge(a, lower, middle);	// Divide array 
            merge(a, middle+1, upper);	// into mid

            combine(a, lower, middle, upper);	// Sort and combine
        }

    }

    // DRIVER CODE
    public static void main(String[] args) {
        
        // Inputs
        int[] array = {23, -6, -3, 7, 34, 9, 66, 23, 55, 1};
        int n = 10;
        
        // Unsorted Array Output
        System.out.println("Unsorted Array :-");
        for (int i=0; i<n; i++)
        System.out.print(array[i]+"   ");

        // Sorting Array
        merge(array, 0, n-1);

        // Sorted Array Output
        System.out.println("\nSorted Array :-");
        for (int i=0; i<n; i++)
            System.out.print(array[i]+"   ");
        System.out.println();

    }
}

File: Python/Sorting/MergeSort.py
# Combine Function
def combine(a, l, m, u):
    i, j, k = l, m+1, 0
    temp = []

    # Compare lower & upper part and store in temp
    while i<=m and j<=u:
        if a[i] <= a[j]:
            temp.append(a[i])
            i += 1
        else:
            temp.append(a[j])
            j += 1
        k += 1

    # Copy remaining lower part
    while i <= m:
        temp.append(a[i])
        i += 1
        k += 1
    
    # Copy remaining upper part
    while j <= u:
        temp.append(a[j])
        j += 1
        k += 1
        
    # Copy sorted temp to main array
    a[l:u+1] = temp


# Merge Sort Function
def merge(a, lower, upper):
    if lower < upper:
        middle = (lower + upper)//2

        merge(a, lower, middle) # Divide array 
        merge(a, middle+1, upper)   # into mid

        combine(a, lower, middle, upper)    # Sort and combine



if __name__ == "__main__":
    # Inputs
    numberList = [23, -6, -3, 7, 34, 9, 66, 23, 55, 1]
    n = len(numberList)

    # Unsorted List Output
    print(numberList)

    # Sorting List
    merge(numberList, 0, n-1)

    # Sorted List Output
    print(numberList)
